plugins/audio-s5gen2/fu-audio-s5gen2-ble-device.c

```diff
@@ -0,0 +1,615 @@
+		return FALSE;
+
+	validate = fu_struct_qc_gaia_v3_set_transport_info_parse(buf, read_len, 0, error);
+	if (validate == NULL)
+		return FALSE;
+
+	return TRUE;
+}
+
+static void
+fu_qc_s5gen2_ble_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(device);
+	g_autofree gchar *api_version;
+
+	fwupd_codec_string_append_size(str, idt, "MTU", self->mtu);
+	fwupd_codec_string_append_hex(str, idt, "GAIAv3 Vendor ID", self->vid_v3);
+	api_version = g_strdup_printf("%u.%u", self->api_version.major, self->api_version.minor);
+	fwupd_codec_string_append(str, idt, "GAIA API version", api_version);
+	fwupd_codec_string_append_hex(str, idt, "GAIA Core feature version", self->feature.core);
+	fwupd_codec_string_append_hex(str, idt, "GAIA DFU feature version", self->feature.dfu);
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_notify_acquire_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(device);
+
+	if (!fu_qc_s5gen2_ble_device_notify_release(self, error))
+		return FALSE;
+
+	return fu_qc_s5gen2_ble_device_notify_acquire(self, error);
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_probe(FuDevice *device, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(device);
+	g_autofree gchar *vendor_id = NULL;
+
+	if (!FU_DEVICE_CLASS(fu_qc_s5gen2_ble_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* after reboot the device might appear too fast */
+	if (!fu_device_retry_full(device,
+				  fu_qc_s5gen2_ble_device_notify_acquire_cb,
+				  FU_QC_S5GEN2_BLE_DEVICE_AQUIRE_RETRIES,
+				  FU_QC_S5GEN2_BLE_DEVICE_TIMEOUT,
+				  NULL,
+				  error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_get_api(self, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_get_features(self, FALSE, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_get_serial(self, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_get_variant(self, error))
+		return FALSE;
+
+	if (self->feature.core >= 2)
+		if (!fu_qc_s5gen2_ble_device_set_transport_protocol(
+			self,
+			FU_QC_S5GEN2_GAIA_PROTOCOL_VERSION,
+			error))
+			return FALSE;
+
+	/* set vendor ID to avoid update error */
+	vendor_id = g_strdup_printf("BLUETOOTH:%04X", self->vid_v3);
+	fu_device_add_vendor_id(device, vendor_id);
+
+	if (!fu_qc_s5gen2_ble_device_register_notification(self, error))
+		return FALSE;
+
+	return fu_qc_s5gen2_ble_device_notify_release(FU_QC_S5GEN2_BLE_DEVICE(device), error);
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_set_quirk_kv(FuDevice *device,
+				     const gchar *key,
+				     const gchar *value,
+				     GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(device);
+	guint64 tmp = 0;
+
+	if (g_strcmp0(key, "AudioS5gen2Gaia3VendorId") == 0) {
+		if (!fu_strtoull(value, &tmp, 0, G_MAXUINT16, FU_INTEGER_BASE_AUTO, error))
+			return FALSE;
+		self->vid_v3 = tmp;
+		return TRUE;
+	}
```

Comment:
rather than doing this, could we use `VendorId = ` in the quirk file? You'd need to do something like `VendorId = BLUETOOTH:0x1234` but I think that's fine.

Reply:
I don't think so.
This is needed for transport protocol, not for device indication, and added as a part of every message.
And that's not the same as VendorID. I.e. Vendor ID could vary, but if it uses the common protocol from QC -- the default value would be used. But any Vendor __could__change it to __any__ own value. Moreover -- that's for v3 protocol, for v2 it is the other value. So, I'd prefer to use plugin-related key here, if you don't mind.

Reply:
Ahh, no problem - got it.